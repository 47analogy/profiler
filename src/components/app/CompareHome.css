.compareHome {
  width: 75%;
  margin: auto;
  padding: 8em;

  border: 1px solid #CCC;
  border-radius: 3px;
  box-shadow: 0 5px 25px #0b1f50;
  background: #fff;
}

.compareHomeForm {
  display: grid;
  grid-template-columns: auto 1fr;
  grid-gap: 1em;
  align-items: center;
}

.compareHomeSubmitButton {
  grid-column-start: span 2
}

.compareHomeFormLabel {
  white-space: nowrap;
}

@media (max-width: 450px) {
  .compareHome {
    width: auto;
    padding: 1em;
  }
}
