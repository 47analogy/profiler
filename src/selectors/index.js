import { createSelector } from 'reselect';
import * as ProfileData from '../profile-data';
import * as ProfileTree from '../profile-tree';

export const getProfileView = state => state.profileView;
export const getProfile = state => getProfileView(state).profile;
export const getProfileInterval = state => getProfile(state).meta.interval;
export const getProfileViewOptions = state => getProfileView(state).viewOptions;
export const getJSOnly = (state, props) => ('jsOnly' in props.location.query);
export const getInvertCallstack = (state, props) => ('invertCallstack' in props.location.query);

export const getScrollToSelectionGeneration = createSelector(
  getProfileViewOptions,
  viewOptions => viewOptions.scrollToSelectionGeneration
);

export const getDisplayRange = createSelector(
  getProfileViewOptions,
  viewOptions => {
    if (viewOptions.rangeFilters.length > 0) {
      return viewOptions.rangeFilters[viewOptions.rangeFilters.length - 1];
    }
    return viewOptions.rootRange;
  }
);

export const getSelectedThreadIndex = createSelector(
  getProfileViewOptions,
  viewOptions => viewOptions.selectedThread
);

export const getZeroAt = createSelector(
  getProfileViewOptions,
  viewOptions => viewOptions.zeroAt
);

export const getThreadOrder = createSelector(
  getProfileViewOptions,
  viewOptions => viewOptions.threadOrder
);

const selectorsForThreads = {};

export const selectorsForThread = threadIndex => {
  if (!(threadIndex in selectorsForThreads)) {
    const getThread = state => getProfile(state).threads[threadIndex];
    const getViewOptions = state => getProfileViewOptions(state).threads[threadIndex];
    const getRangeFilteredThread = createSelector(
      getThread,
      getDisplayRange,
      (thread, range) => {
        const { start, end } = range;
        return ProfileData.filterThreadToRange(thread, start, end);
      }
    );
    const getJSOnlyFilteredThread = createSelector(
      getRangeFilteredThread,
      getJSOnly,
      (thread, jsOnly) => {
        return jsOnly ? ProfileData.filterThreadToJSOnly(thread) : thread;
      }
    );
    const getFilteredThread = createSelector(
      getJSOnlyFilteredThread,
      getInvertCallstack,
      (thread, shouldInvertCallstack) => {
        return shouldInvertCallstack ? ProfileData.invertCallstack(thread) : thread;
      }
    );
    const getRangeSelectionFilteredThread = createSelector(
      getFilteredThread,
      getProfileViewOptions,
      (thread, viewOptions) => {
        if (!viewOptions.selection.hasSelection) {
          return thread;
        }
        const { selectionStart, selectionEnd } = viewOptions.selection;
        return ProfileData.filterThreadToRange(thread, selectionStart, selectionEnd);
      }
    );
    const getFuncStackInfo = createSelector(
      getFilteredThread,
      ({stackTable, frameTable, funcTable}) => {
        return ProfileData.getFuncStackInfo(stackTable, frameTable, funcTable);
      }
    );
    const getSelectedFuncStackAsFuncArray = createSelector(
      getViewOptions,
      threadViewOptions => threadViewOptions.selectedFuncStack
    );
    const getSelectedFuncStack = createSelector(
      getFuncStackInfo,
      getSelectedFuncStackAsFuncArray,
      (funcStackInfo, funcArray) => {
        return ProfileData.getFuncStackFromFuncArray(funcArray, funcStackInfo.funcStackTable);
      }
    );
    const getExpandedFuncStacksAsFuncArrays = createSelector(
      getViewOptions,
      threadViewOptions => threadViewOptions.expandedFuncStacks
    );
    const getExpandedFuncStacks = createSelector(
      getFuncStackInfo,
      getExpandedFuncStacksAsFuncArrays,
      (funcStackInfo, funcArrays) => {
        return funcArrays.map(funcArray => ProfileData.getFuncStackFromFuncArray(funcArray, funcStackInfo.funcStackTable));
      }
    );
    const getCallTree = createSelector(
      getRangeSelectionFilteredThread,
      getProfileInterval,
      getFuncStackInfo,
      ProfileTree.getCallTree
    );
    selectorsForThreads[threadIndex] = {
      getThread,
      getViewOptions,
      getFilteredThread,
      getRangeSelectionFilteredThread,
      getFuncStackInfo,
      getSelectedFuncStack,
      getExpandedFuncStacks,
      getCallTree,
    };
  }
  return selectorsForThreads[threadIndex];
};

export const selectedThreadSelectors = (() => {
  const anyThreadSelectors = selectorsForThread(0);
  const result = {};
  for (const key in anyThreadSelectors) {
    result[key] = (state, props) => selectorsForThread(getSelectedThreadIndex(state, props))[key](state, props);
  }
  return result;
})();
