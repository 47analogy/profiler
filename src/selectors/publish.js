/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { createSelector } from 'reselect';
import {
  getProfile,
  getProfileRootRange,
  getCommittedRange,
  getGlobalTracks,
  getLocalTracksByPid,
} from './profile';
import { compress } from '../utils/gz';
import {
  serializeProfile,
  sanitizePII,
} from '../profile-logic/process-profile';
import prettyBytes from '../utils/pretty-bytes';
import { getHiddenGlobalTracks, getHiddenLocalTracksByPid } from './url-state';
import { ensureExists } from '../utils/flow';
import { formatNumber } from '../utils/format-numbers';

import type { PublishState, UploadState, UploadPhase } from '../types/state';
import type { Selector } from '../types/store';
import type { CheckedSharingOptions } from '../types/actions';
import type { RemoveProfileInformation } from '../types/profile-derived';

export const getPublishState: Selector<PublishState> = state => state.publish;

export const getCheckedSharingOptions: Selector<
  CheckedSharingOptions
> = state => getPublishState(state).checkedSharingOptions;

export const getFilenameString: Selector<string> = createSelector(
  getProfile,
  getProfileRootRange,
  (profile, rootRange) => {
    const { startTime, product } = profile.meta;

    // Pad single digit numbers with a 0.
    const pad = x => (x < 10 ? `0${x}` : `${x}`);

    // Compute the date string.
    const date = new Date(startTime + rootRange.start);
    const year = pad(date.getFullYear());
    const month = pad(date.getMonth() + 1);
    const day = pad(date.getDate());
    const hour = pad(date.getHours());
    const min = pad(date.getMinutes());
    const dateString = `${year}-${month}-${day} ${hour}.${min}`;

    // Return the final file name
    return `${product} ${dateString} profile.json`;
  }
);

export const getRemoveProfileInformation: Selector<RemoveProfileInformation | null> = createSelector(
  getCheckedSharingOptions,
  getProfile,
  getCommittedRange,
  getHiddenGlobalTracks,
  getHiddenLocalTracksByPid,
  getGlobalTracks,
  getLocalTracksByPid,
  (
    checkedSharingOptions,
    profile,
    committedRange,
    hiddenGlobalTracks,
    hiddenLocalTracksByPid,
    globalTracks,
    localTracksByPid
  ) => {
    if (!checkedSharingOptions.isFiltering) {
      return null;
    }

    // Find all of the thread indexes that are hidden.
    const shouldRemoveThreads = new Set();
    if (checkedSharingOptions.hiddenThreads) {
      for (const globalTrackIndex of hiddenGlobalTracks) {
        const globalTrack = globalTracks[globalTrackIndex];
        if (
          globalTrack.type === 'process' &&
          globalTrack.mainThreadIndex !== null
        ) {
          // This is a process thread that has been hidden.
          shouldRemoveThreads.add(globalTrack.mainThreadIndex);
          const localTracks = ensureExists(
            localTracksByPid.get(globalTrack.pid),
            'Expected to be able to get a local track by PID.'
          );

          // Also add all of the children threads, as they are hidden as well.
          for (const localTrack of localTracks) {
            if (localTrack.type === 'thread') {
              shouldRemoveThreads.add(localTrack.threadIndex);
            }
          }
        }
      }

      // Add all of the local tracks that have been hidden.
      for (const [pid, hiddenLocalTrackIndexes] of hiddenLocalTracksByPid) {
        const localTracks = ensureExists(
          localTracksByPid.get(pid),
          'Expected to be able to get a local track by PID'
        );
        for (const hiddenLocalTrackIndex of hiddenLocalTrackIndexes) {
          const localTrack = localTracks[hiddenLocalTrackIndex];
          if (localTrack.type === 'thread') {
            shouldRemoveThreads.add(localTrack.threadIndex);
          }
        }
      }
    }

    return {
      shouldFilterToCommittedRange: checkedSharingOptions.timeRange
        ? committedRange
        : null,
      shouldRemoveNetworkUrls: checkedSharingOptions.urls,
      shouldRemoveAllUrls: checkedSharingOptions.urls,
      shouldRemoveThreadsWithScreenshots: new Set(
        checkedSharingOptions.screenshots
          ? profile.threads.map((_, threadIndex) => threadIndex)
          : []
      ),
      shouldRemoveThreads,
      shouldRemoveExtensions: checkedSharingOptions.extension,
    };
  }
);

export const getSanitizedProfileBlob: Selector<Promise<Blob>> = createSelector(
  getRemoveProfileInformation,
  getProfile,
  async (removeProfileInformation, profile) => {
    const maybeSanitizedProfile = removeProfileInformation
      ? sanitizePII(profile, removeProfileInformation)
      : profile;
    const data = await compress(serializeProfile(maybeSanitizedProfile));
    return new Blob([data], { type: 'application/octet-binary' });
  }
);

export const getCompressedProfileBlobUrl: Selector<
  Promise<string>
> = createSelector(getSanitizedProfileBlob, blobPromise =>
  blobPromise.then(blob => URL.createObjectURL(blob))
);

export const getDownloadSize: Selector<Promise<string>> = createSelector(
  getSanitizedProfileBlob,
  blobPromise => blobPromise.then(blob => prettyBytes(blob.size))
);

/**
 * In order to use React keyed components for a PII filtered profile, we need
 * an easy string or number to use for the key that represents the current version. This
 * selector creates a generation value that increases every time the remove profile
 * information or profile changes.
 */
let _sanitizedProfileGeneration = 0;
export const getSanitizedProfileGeneration: Selector<number> = createSelector(
  getRemoveProfileInformation,
  getProfile,
  () => _sanitizedProfileGeneration++
);

export const getUploadState: Selector<UploadState> = state =>
  getPublishState(state).upload;

export const getUploadPhase: Selector<UploadPhase> = state =>
  getUploadState(state).phase;

export const getUploadGeneration: Selector<number> = state =>
  getUploadState(state).generation;

export const getUploadProgress: Selector<number> = state =>
  getUploadState(state).uploadProgress;

export const getUploadUrl: Selector<string> = state =>
  getUploadState(state).url;

export const getUploadError: Selector<Error | mixed> = state =>
  getUploadState(state).error;

export const getUploadProgressString: Selector<string> = createSelector(
  getUploadProgress,
  progress => formatNumber(progress, 0, 0, 'percent')
);

export const getAbortFunction: Selector<() => void> = state =>
  getUploadState(state).abortFunction;
