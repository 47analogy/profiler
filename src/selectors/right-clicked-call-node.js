/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { createSelector } from 'reselect';
import { getCallNodeIndexFromPath } from '../profile-logic/profile-data';
import { getThreadSelectors } from './per-thread';
import { getProfileViewOptions } from './profile';

import type { Thread, ThreadIndex } from '../types/profile';
import type {
  IndexIntoCallNodeTable,
  CallNodePath,
  CallNodeInfo,
} from '../types/profile-derived';
import type { Selector } from '../types/store';

export type RightClickedCallNodeInfo = {|
  +threadIndex: ThreadIndex,
  +thread: Thread,
  +callNodePath: CallNodePath,
  +callNodeIndex: IndexIntoCallNodeTable,
  +callNodeInfo: CallNodeInfo,
|};

export const getRightClickedCallNodeInfo: Selector<RightClickedCallNodeInfo | null> = createSelector(
  state => state,
  state => {
    const { rightClickedCallNode } = getProfileViewOptions(state);

    if (rightClickedCallNode === null) {
      return null;
    }

    const { getFilteredThread, getCallNodeInfo } = getThreadSelectors(
      rightClickedCallNode.threadIndex
    );

    const thread = getFilteredThread(state);

    const callNodeInfo = getCallNodeInfo(state);

    const callNodeIndex = getCallNodeIndexFromPath(
      rightClickedCallNode.callNodePath,
      callNodeInfo.callNodeTable
    );

    if (callNodeIndex === null) {
      return null;
    }

    return {
      ...rightClickedCallNode,
      thread,
      callNodeIndex,
      callNodeInfo,
    };
  }
);
