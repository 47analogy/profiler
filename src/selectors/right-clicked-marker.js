/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { createSelector } from 'reselect';
import { getThreadSelectors } from './per-thread';
import { getProfileViewOptions } from './profile';

import type { ThreadIndex } from '../types/profile';
import type { MarkerIndex, Marker } from '../types/profile-derived';
import type { Selector } from '../types/store';

export const getRightClickedMarkerThreadIndex: Selector<ThreadIndex | null> = createSelector(
  getProfileViewOptions,
  ({ rightClickedMarker }) =>
    rightClickedMarker ? rightClickedMarker.threadIndex : null
);

export const getRightClickedMarkerIndex: Selector<MarkerIndex | null> = createSelector(
  getProfileViewOptions,
  ({ rightClickedMarker }) =>
    rightClickedMarker ? rightClickedMarker.markerIndex : null
);

export const getRightClickedMarker: Selector<Marker | null> = createSelector(
  getRightClickedMarkerIndex,
  state => {
    const threadIndex = getRightClickedMarkerThreadIndex(state);
    if (threadIndex !== null) {
      return getThreadSelectors(threadIndex).getMarkerGetter(state);
    }

    return null;
  },
  (markerIndex, getMarker) => {
    if (markerIndex !== null && getMarker !== null) {
      return getMarker(markerIndex);
    }

    return null;
  }
);
