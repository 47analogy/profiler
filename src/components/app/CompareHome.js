/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React, { PureComponent } from 'react';
import { changeProfilesToCompare } from '../../actions/app';

import explicitConnect from '../../utils/connect';
import type {
  ExplicitConnectOptions,
  ConnectedProps,
} from '../../utils/connect';

import './CompareHome.css';

type DispatchProps = {|
  +changeProfilesToCompare: typeof changeProfilesToCompare,
|};

type Props = ConnectedProps<{||}, {||}, DispatchProps>;

type State = {
  profile1: string,
  profile2: string,
};

class CompareHome extends PureComponent<Props, State> {
  state = { profile1: '', profile2: '' };

  handleInputChange = (event: SyntheticInputEvent<>) => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };

  handleFormSubmit = (e: SyntheticEvent<>) => {
    e.preventDefault();
    let { profile1, profile2 } = this.state;
    const { changeProfilesToCompare } = this.props;
    if (!profile1.startsWith('http')) {
      profile1 = 'http://' + profile1;
    }
    if (!profile2.startsWith('http')) {
      profile2 = 'http://' + profile2;
    }
    changeProfilesToCompare(profile1, profile2);
  };

  render() {
    const { profile1, profile2 } = this.state;

    return (
      <form className="compareHome" onSubmit={this.handleFormSubmit}>
        <h1>Enter the 2 profile URLs that youâ€™d like to compare</h1>
        <p>
          The tool will extract the data from the selected track and range for
          each profile, and put them both on the same view to make them easy to
          compare.
        </p>
        <label htmlFor="compareHomeProfile1">Profile 1:</label>
        <input
          name="profile1"
          id="compareHomeProfile1"
          className="compareHome-input"
          onChange={this.handleInputChange}
          value={profile1}
        />
        <label htmlFor="compareHomeProfile2">Profile 2:</label>
        <input
          name="profile2"
          id="compareHomeProfile2"
          className="compareHome-input"
          onChange={this.handleInputChange}
          value={profile2}
        />
        <input
          className="compareHome-submit-button"
          type="submit"
          value="Retrieve profiles"
        />
      </form>
    );
  }
}

const options: ExplicitConnectOptions<{||}, {||}, Props> = {
  mapDispatchToProps: { changeProfilesToCompare },
  component: CompareHome,
};

export default explicitConnect(options);
