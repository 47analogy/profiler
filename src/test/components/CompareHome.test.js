/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import React from 'react';
import { Provider } from 'react-redux';
import { render, fireEvent, cleanup } from 'react-testing-library';

import CompareHome from '../../components/app/CompareHome';
import { getProfilesToCompare } from '../../selectors/url-state';

import { blankStore } from '../fixtures/stores';

describe('app/CompareHome', () => {
  afterEach(cleanup);

  it('renders properly', () => {
    const store = blankStore();
    const { container, getByLabelText, getByText } = render(
      <Provider store={store}>
        <CompareHome />
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();

    fireEvent.change(getByLabelText(/Profile 1/), {
      target: { value: 'www.url11.com' },
    });
    fireEvent.change(getByLabelText(/Profile 2/), {
      target: { value: 'www.url12.com' },
    });
    const retrieveButton = getByText(
      (content, element) =>
        element instanceof HTMLInputElement && /Retrieve/.test(element.value)
    );
    fireEvent.click(retrieveButton);

    expect(getProfilesToCompare(store.getState())).toEqual([
      'http://www.url11.com',
      'http://www.url12.com',
    ]);
  });
});
