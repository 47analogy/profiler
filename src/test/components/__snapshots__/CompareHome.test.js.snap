// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/CompareHome renders properly 1`] = `
<form
  class="compareHome"
>
  <h1>
    Enter the 2 profile URLs that youâ€™d like to compare
  </h1>
  <p>
    The tool will extract the data from the selected track and range for each profile, and put them both on the same view to make them easy to compare.
  </p>
  <label
    for="compareHomeProfile1"
  >
    Profile 1:
  </label>
  <input
    class="compareHome-input"
    id="compareHomeProfile1"
    name="profile1"
    value=""
  />
  <label
    for="compareHomeProfile2"
  >
    Profile 2:
  </label>
  <input
    class="compareHome-input"
    id="compareHomeProfile2"
    name="profile2"
    value=""
  />
  <input
    class="compareHome-submit-button"
    type="submit"
    value="Retrieve profiles"
  />
</form>
`;
